Respect balenaRootCA system-wide

mtoman opened the pull request:
We allow the user to specify a custom CA in the .balenaRootCA key
of config.json but at this moment this is only respected by
the supervisor. This commit adds it to the system-wide CA bundle
so that the CA is respected everywhere.

Fixes #1398

Change-type: minor
Signed-off-by: Michal Toman <michalt@balena.io>


---
### Contributor checklist
<!-- For completed items, change [ ] to [x].  -->
- [x] Changes have been tested
- [x] `Change-type` present on at least one commit
- [x] `Signed-off-by` is present
- [x] The PR complies with the [Open Embedded Commit Patch Message Guidelines](http://www.openembedded.org/wiki/Commit_Patch_Message_Guidelines)
<!-- optional: `Changelog-entry` present on at least one commit if you want to set the changelog entry manually-->

### Reviewer Guidelines
- When submitting a review, please pick:
  - '*Approve*' if this change would be acceptable in the codebase (even if there are minor or cosmetic tweaks that could be improved).
  - '*Request Changes*' if this change would not be acceptable in our codebase (e.g. bugs, changes that will make development harder in future, security/performance issues, etc).
  - '*Comment*' if you don't feel you have enough information to decide either way (e.g. if you have major questions, or you don't understand the context of the change sufficiently to fully review yourself, but want to make a comment)

mtoman commented:
This could use some real world setup testing, I have only performed artificial tests.

alexgg commented on meta-balena-common/recipes-support/ca-certificates/ca-certificates/extract-balena-ca:
```diff
@@ -0,0 +1,46 @@
+#!/bin/sh
+
+CONFIG_JSON="/mnt/boot/config.json"
+CA_BUNDLE="/etc/ssl/certs/ca-certificates.crt"
```
CA_BUNDLE is unused? Good habit to run shellcheck on shell scripts :)

mtoman replied:
Damn I did changes after my "final" check :) Will remove that.

alexgg commented on meta-balena-common/recipes-support/ca-certificates/ca-certificates/extract-balena-ca:
```diff
@@ -0,0 +1,46 @@
+JSON_CA_KEY=".balenaRootCA"
+BALENA_CA_FILE="/usr/share/ca-certificates/balena/balenaRootCA.crt"
+
+log () {
```
Consider having homogeneous logging by sourcing  `os-helpers-logging`. It's only used in the initramfs at the moment, but it may be good to start using it in new shell scripts so that the output looks the same for all.

mtoman replied:
+1 I was not aware of that. `os-helpers` are not installed in the OS right now, are you OK with pulling it in with this PR?

alexgg replied:
I think it's a good moment. It won't add a lot of space and it will help to make the output of script homogeneous.

alexgg commented on meta-balena-common/recipes-support/ca-certificates/ca-certificates/extract-balena-ca:
```diff
@@ -0,0 +1,46 @@
+log "Adding the custom CA from config.json"
+
+# Copy new CA to its place
+cp -a "${DEST_FILE}" "${BALENA_CA_FILE}"
```
If you mv here you don't need to rm later.

mtoman replied:
I do not like cross-fs moves but that might be my personal paranoia, if you prefer `mv` I will change that.

alexgg replied:
it's a nitpick, leave it as is.

mtoman replied:
:+1:

alexgg commented on meta-balena-common/recipes-support/ca-certificates/ca-certificates/extract-balena-ca.service:
```diff
@@ -0,0 +1,12 @@
+[Unit]
```
Wondering whether this unit should run whenever the config.json custom CA is updated.
I guess the supervisor will reboot the unit after modifying config.json anyway, so it's good as is. But conceptually I don't know why the reboot is needed in all cases when changes could be applied on the fly. I will ping the supervisor team to understand this a bit more.

mtoman replied:
Discussed offline, followup below.

alexgg commented on meta-balena-common/recipes-support/ca-certificates/ca-certificates/extract-balena-ca:
```diff
@@ -0,0 +1,46 @@
+cp -a "${DEST_FILE}" "${BALENA_CA_FILE}"
+
+# Rebuild the CA bundle - for this we need to make /etc/ssl/certs writable
+systemctl start bind-etc-ssl-certs.service
```
Maybe check whether it's already run.

mtoman replied:
Isn't systemd taking care of that? I was always assuming `systemctl start` means "start if not started" but reading through docs now that is not mentioned anywhere so I might be wrong.

alexgg replied:
Well, if it ran an exited this may re-run it, probably depending on whether `RemainAfterExit=` is set or not. It may be harmless anyway but worth checking.

alexgg commented:
@mtoman see the conversation in https://www.flowdock.com/app/rulemotion/r-supervisor/threads/UhrL-JCTcLgKv5GoinP-REoa5fG. This change will need to apply on the fly, and the supervisor will also need to be restarted as it uses the root CA.

alexgg commented:
@mtoman Some tests we should do before merge include:

- HUP scripts
https://github.com/balena-os/balenahup/issues/213

- redsocks proxy to use custom CAs
https://jel.ly.fish/#/support-thread~bc290f19-dca9-4eb2-90b8-a36f9f7cd99a

- curl, for resin-device-progress and update-resin-supervisor
test update-resin-supervisor in BoB

And is the addition of multiple certificates also working?

mtoman commented:
Addressed the suggestions and tested (on a Pi3/32 for the record, though the device type should be irrelevant).
Adding multiple certificates spits a warning from `update-ca-certificates` but works.
If the change is happening on the fly, the supervisor is the one triggering it. That is why the service does not restart the supervisor but the supervisor should start `extract-balena-ca.service`. I have removed the `RemainAfterExit=yes` because of this as it is now not a run-once thing.
Fixed a service ordering bug that made `/etc/ssl/certs` randomly (not) mounted after a HUP.

mtoman commented:
This obsoletes two mentions in the supervisor:
https://github.com/balena-io/balena-supervisor/blob/master/entry.sh#L34
https://github.com/balena-io/balena-supervisor/blob/master/tools/dind/start-resin-supervisor#L33
These are not conflicting in any way but can be safely removed.

This also obsoletes handling in os-config
https://github.com/balena-os/os-config/blob/master/src/config_json.rs#L74

Both should be fine to remove but need testing.

alexgg commented:
> This obsoletes two mentions in the supervisor:
> https://github.com/balena-io/balena-supervisor/blob/master/entry.sh#L34
> https://github.com/balena-io/balena-supervisor/blob/master/tools/dind/start-resin-supervisor#L33
> These are not conflicting in any way but can be safely removed.
> 
> This also obsoletes handling in os-config
> https://github.com/balena-os/os-config/blob/master/src/config_json.rs#L74
> 
> Both should be fine to remove but need testing.

@mtoman Maybe add the corresponding issues in both the supervisor and os-config repositories

alexgg commented:
@balena-ci retest

alexgg commented:
@balena-ci retest

alexgg commented:
> > This obsoletes two mentions in the supervisor:
> > https://github.com/balena-io/balena-supervisor/blob/master/entry.sh#L34
> > https://github.com/balena-io/balena-supervisor/blob/master/tools/dind/start-resin-supervisor#L33
> > These are not conflicting in any way but can be safely removed.
> > This also obsoletes handling in os-config
> > https://github.com/balena-os/os-config/blob/master/src/config_json.rs#L74
> > Both should be fine to remove but need testing.
> 
> @mtoman Maybe add the corresponding issues in both the supervisor and os-config repositories

I created:

https://github.com/balena-os/os-config/issues/30
https://github.com/balena-io/balena-supervisor/issues/1444

alexgg commented:
@balena-ci retest

alexgg commented:
@resin-jenkins retest this please

alexgg commented:
@balena-ci retest

alexgg commented:
@balena-ci rebase

alexgg commented:
@balena-ci retest this please

alexgg commented:
@balena-ci test

alexgg commented:
@balena-ci test

alexgg commented:
@balena-ci test

alexgg commented:
@resin-jenkins retest this please
